mpmath/libmp/libmpf.py

```diff
@@ -1101,6 +1102,49 @@ def to_digits_exp(s, dps, base=10):
     exponent += len(digits) - fixdps - 1
     return sign, digits, exponent
 
+def round_digits(digits, dps, base):
+    '''
+    Returns the rounded digits, and the number of places the decimal point was
+    shifted.
+    '''
+
+    exponent = 0
+
+    rnd_digs = stddigits[(base//2 + base % 2):base]
+
+    round_up = False
+    # The first digit after dps is a 5.
+    if digits[dps] == rnd_digs[0]:
+        for i in range(dps+1, len(digits)):
+            if digits[i] != '0':
+                round_up = True
+                break
+        if digits[dps-1] in stddigits[1:base:2]:
+            round_up = True
+
+    if not round_up:
+        digits = digits[:dps] + stddigits[int(digits[dps], base) - 1]
+
+    # Rounding up kills some instances of "...99999"
+    if len(digits) > dps and digits[dps] in rnd_digs:
```

Comment:
I am OK with the changes. The only small caveat is that now `to_str` rounds to nearest, ties away from zero whereas `format_mpf` rounds to nearest, ties to even. I am OK with this as well.

I removed the check for `len(digits)` because `digits` always comes from `to_digits_exp`, called with `dps = dps+3`, meaning that `digits` has always more than `dps` numbers.

In the case of the new `round_digits` function, since this check was on a line by itself, the branch where `len(digits) <= dps` was never covered by the tests, so I had to remove the line to pass the coverage test. I could not think of a way to test for this branch without resorting to calls to functions that are not meant to be used by the end user.

Reply:
The rounding behaviour of to_str() seems to be a feature.  We might consider changing that, but it will costs a backward compatibility.

> I removed the check for len(digits) because digits always comes from to_digits_exp, called with dps = dps+3, meaning that digits has always more than dps numbers.

Maybe you should add a comment or assert.